test/unit/location_test.rb

```diff
@@ -3,133 +3,214 @@
 class LocationTest < ActiveSupport::TestCase
   include ActiveShipping::Test::Fixtures
 
-  def test_countries
-    assert_instance_of ActiveUtils::Country, location_fixtures[:ottawa].country
-    assert_equal 'CA', location_fixtures[:ottawa].country_code(:alpha2)
-  end
-
-  def test_location_from_strange_hash
-    hash = {  :country => 'CA',
-              :zip => '90210',
-              :territory_code => 'QC',
-              :town => 'Perth',
-              :address => '66 Gregory Ave.',
-              :phone => '515-555-1212',
-              :fax_number => 'none to speak of',
-              :address_type => :commercial
-            }
-    location = Location.from(hash)
-
-    assert_equal hash[:country], location.country_code(:alpha2)
-    assert_equal hash[:zip], location.zip
-    assert_equal hash[:territory_code], location.province
-    assert_equal hash[:town], location.city
-    assert_equal hash[:address], location.address1
-    assert_equal hash[:phone], location.phone
-    assert_equal hash[:fax_number], location.fax
-    assert_equal hash[:address_type].to_s, location.address_type
-  end
-
-  def test_pretty_print
-    expected = "110 Laurier Avenue West\nOttawa, ON, K1P 1J1\nCanada"
-    assert_equal expected, location_fixtures[:ottawa].prettyprint
+  setup do
+    @location = location_fixtures[:ottawa]
+    @address2 = 'Apt 613'
+    @address3 = 'Victory Lane'
+    @attributes_hash = {
+      country: 'CA',
+      zip: '90210',
+      territory_code: 'QC',
+      town: 'Perth',
+      address: '66 Gregory Ave.',
+      phone: '515-555-1212',
+      fax_number: 'none to speak of',
+      address_type: :commercial,
+      name: "Bob Bobsen",
+    }
+    # @attributes_object = OpenStruct.new(@attributes_hash)
+  end
+
+  test "#initialize sets a country object" do
+    assert_instance_of ActiveUtils::Country, @location.country
+    assert_equal 'CA', @location.country_code(:alpha2)
+  end
+
+  test ".from sets up the location from a hash" do
+    location = Location.from(@attributes_hash)
+
+    assert_equal @attributes_hash[:country], location.country_code(:alpha2)
+    assert_equal @attributes_hash[:zip], location.zip
+    assert_equal @attributes_hash[:territory_code], location.province
+    assert_equal @attributes_hash[:town], location.city
+    assert_equal @attributes_hash[:address], location.address1
+    assert_equal @attributes_hash[:phone], location.phone
+    assert_equal @attributes_hash[:fax_number], location.fax
+    assert_equal @attributes_hash[:address_type].to_s, location.address_type
+    assert_equal @attributes_hash[:name], location.name
+  end
+
+  test ".from sets from an object with properties" do
```

Comment:
We didn't actually have any tests for `.from` an object, not a hash.

This `Class.new` is a little nasty, but I tried with `OpenStruct` and it still responds to `:[]` so doesn't work.